### DISCUSSION TITLE
IAM auth using CredentialsProvider for ClusterClient

### DISCUSSION THREAD

**Discussion by ParthKulkarni** [redis/go-redis#2710] [Q&A] [answered]
With the recent addition of [support for IAM auth](https://aws.amazon.com/about-aws/whats-new/2022/11/amazon-elasticache-supports-iam-authentication-redis-clusters/) for redis clusters, I'm looking to move away from the password usage and migrate to using IAM auth. I am using `redis.NewClusterClient` to interface with AWS elasticache for Redis. I am authenticating into the redis instance like so,

```
import "github.com/go-redis/redis/v8"

func InitRedis() {
    RedisClient := redis.NewClusterClient(redis.ClusterOptions{
        Addrs:     []string{"redis-address"},
        Password:  "redis-password",
        PoolSize:  10,
    })

    _, err := RedisClient.Ping().Result()
    if err != nil {
        log.Error(err, "could not establish connection")
        return
    }
}
```

I saw [this example](https://github.com/redis/go-redis/discussions/2343) on authenticating using `CredentialsProvider` for a redis `Client`. However, I cannot see how I can do the same for a `ClusterClient` since the [ClusterOptions struct](https://pkg.go.dev/github.com/go-redis/redis/v9#ClusterOptions) does not seem to have support for CredentialsProvider. How should I go about authenticating into elasticache redis using IAM through CredentialsProvider  for a `ClusterClient` with the go-redis package? 

Could we do something like here - [noise.getoto.net/author/younggu-yun](https://noise.getoto.net/author/younggu-yun/) (Search for CredentialsProvider on the page)

**Comment by ndyakov** [comment, marked as answer]
Please check with `go-redis` `v9.8.0` when we do an official release.
